components/eamxx/src/physics/mam/mam_aerosol_optics_read_tables.hpp

```diff
@@ -0,0 +1,366 @@
+  aerosol_optics_host_data.refindex_real_sw_host = refindex_real_sw_host;
+  aerosol_optics_host_data.refindex_im_sw_host   = refindex_im_sw_host;
+  aerosol_optics_host_data.absplw_host           = absplw_host;
+  aerosol_optics_host_data.asmpsw_host           = asmpsw_host;
+  aerosol_optics_host_data.extpsw_host           = extpsw_host;
+  aerosol_optics_host_data.abspsw_host           = abspsw_host;
+
+  FieldLayout scalar_refindex_real_lw_layout{{LWBND, NREFINDEX_REAL},
+                                             {nlwbands, refindex_real}};
+
+  FieldLayout scalar_refindex_im_lw_layout{{LWBND, NREFINDEX_IM},
+                                           {nlwbands, refindex_im}};
+  FieldLayout scalar_refindex_real_sw_layout{{SWBND, NREFINDEX_REAL},
+                                             {nswbands, refindex_real}};
+
+  FieldLayout scalar_refindex_im_sw_layout{{SWBND, NREFINDEX_IM},
+                                           {nswbands, refindex_im}};
+
+  FieldLayout scalar_absplw_layout{
+      {LWBND, MODE, NREFINDEX_IM, NREFINDEX_REAL, NCOEF_NUMBER},
+      {nlwbands, 1, refindex_im, refindex_real, coef_number}};
+  // use also for extpsw, abspsw
+  FieldLayout scalar_asmpsw_layout{
+      {SWBND, MODE, NREFINDEX_IM, NREFINDEX_REAL, NCOEF_NUMBER},
+      {nswbands, 1, refindex_im, refindex_real, coef_number}};
+
+  rrtmg_params.set<strvec_t>(
+      "Field Names",
+      {"asmpsw", "extpsw", "abspsw", "absplw", "refindex_real_sw",
+       "refindex_im_sw", "refindex_real_lw", "refindex_im_lw"});
+
+  rrtmg_params.set("Skip_Grid_Checks", true);
+
+  host_views["refindex_real_sw"] =
+      view_1d_host(refindex_real_sw_host.data(), refindex_real_sw_host.size());
+
+  host_views["refindex_im_sw"] =
+      view_1d_host(refindex_im_sw_host.data(), refindex_im_sw_host.size());
+
+  host_views["refindex_real_lw"] =
+      view_1d_host(refindex_real_lw_host.data(), refindex_real_lw_host.size());
+
+  host_views["refindex_im_lw"] =
+      view_1d_host(refindex_im_lw_host.data(), refindex_im_lw_host.size());
+
+  host_views["absplw"] = view_1d_host(absplw_host.data(), absplw_host.size());
+
+  host_views["asmpsw"] = view_1d_host(asmpsw_host.data(), asmpsw_host.size());
+
+  host_views["extpsw"] = view_1d_host(extpsw_host.data(), extpsw_host.size());
+
+  host_views["abspsw"] = view_1d_host(abspsw_host.data(), abspsw_host.size());
+
+  layouts.emplace("refindex_real_lw", scalar_refindex_real_lw_layout);
+  layouts.emplace("refindex_im_lw", scalar_refindex_im_lw_layout);
+  layouts.emplace("refindex_real_sw", scalar_refindex_real_sw_layout);
+  layouts.emplace("refindex_im_sw", scalar_refindex_im_sw_layout);
+  layouts.emplace("absplw", scalar_absplw_layout);
+  layouts.emplace("asmpsw", scalar_asmpsw_layout);
+  layouts.emplace("extpsw", scalar_asmpsw_layout);
+  layouts.emplace("abspsw", scalar_asmpsw_layout);
+}
+// KOKKOS_INLINE_FUNCTION
+inline void read_rrtmg_table(
```

Comment:
I mentioned the issue of band ordering in a previous comment, but probably important to repeat here. These functions appear to be designed to read the data prepared for RRTMG, not RRTMGP, but I don't see a check to make sure that the band ordering is correct, or to re-order the bands to be correct. Note that RRTMGP has a different band ordering in the SW than RRTMG (the ordering in RRTMG is not monotonic), so in the EAM interface to RRTMGP, we took the lazy way out and read in the RRTMG tables to avoid having to make new tables, but then re-order the optics in the RRTMGP interface. Do we need to assume RRTMG band-ordering for these fields, and take responsibility in RRTMGP to re-order the optics? Or should MAMxx be responsible for providing the correct band ordering? I'm not sure what the better approach would be, but we need to be consistent and make sure we are doing this right!

Reply:
Yes, these functions are design to read netcdf from the mam4, which uses RRTMG. We are doing the re-order in the mam4xx/emaxx optics interface.